### PR TITLE
Clean Up Comments in LCM(-LoRA) Distillation Scripts.

### PR DESCRIPTION
**Pull request by dg845**
# What does this PR do?

This PR cleans up and improves the comments in the LCM and LCM-LoRA distillation scripts. Includes a few small bug fixes.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [x] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [ ] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@patrickvonplaten
@patil-suraj
@sayakpaul

### CODE REVIEW

File: examples/consistency_distillation/train_lcm_distill_lora_sdxl_wds.py [hunk 1]
```diff
@@ -1283,13 +1322,26 @@ def compute_embeddings(
...
+                            alpha_schedule,
+                            sigma_schedule,
+                        )
 
-                        # 20.4.11. Perform "CFG" to get x_prev estimate (using the LCM paper's CFG formulation)
+                        # 3. Calculate the CFG estimate of x_0 (pred_x0) and eps_0 (pred_noise)
+                        # Note that this uses the LCM paper's CFG formulation rather than the Imagen CFG formulation
                         pred_x0 = cond_pred_x0 + w * (cond_pred_x0 - uncond_pred_x0)
-                        pred_noise = cond_teacher_output + w * (cond_teacher_output - uncond_teacher_output)
+                        pred_noise = cond_pred_noise + w * (cond_pred_noise - uncond_pred_noise)
```
Comment (sayakpaul) [inline examples/consistency_distillation/train_lcm_distill_lora_sdxl_wds.py:1337]: Hmm this seems like a big change to me. I wonder how much of an effect it has on the results.

Comment (dg845) [reply examples/consistency_distillation/train_lcm_distill_lora_sdxl_wds.py:1337]: Hi @sayakpaul, this change should be a no-op when the teacher LDM checkpoint has `self.scheduler.config.prediction_type == "epsilon"` (as is the case for the [`runwayml/stable-diffusion-v1-5`](https://huggingface.co/runwayml/stable-diffusion-v1-5/blob/main/scheduler/scheduler_config.json) and [`stabilityai/stable-diffusion-xl-base-1.0`](https://huggingface.co/stabilityai/stable-diffusion-xl-base-1.0/blob/main/scheduler/scheduler_config.json) checkpoints suggested in the `README` examples). The current logic of directly using the residual output of the teacher `unet` is correct when the `prediction_type` is `epsilon`:

https://github.com/huggingface/diffusers/blob/4836cfad9836e6742a1d09462f85313534388a48/examples/consistency_distillation/train_lcm_distill_lora_sdxl_wds.py#L1289

but in general we need to convert the residual `unet` output to a predicted epsilon for use in `DDIMSolver.ddim_step`, which is done by the new `predicted_source_noise` function, which is based on `DDIMScheduler.step`:

https://github.com/huggingface/diffusers/blob/4836cfad9836e6742a1d09462f85313534388a48/src/diffusers/schedulers/scheduling_ddim.py#L412-L422

Note that `pred_epsilon` is exactly equal to `model_output` when `prediction_type == "epsilon"`.

Comment (sayakpaul) [reply examples/consistency_distillation/train_lcm_distill_lora_sdxl_wds.py:1337]: Thanks for the solid explanation!

**Review by sayakpaul — APPROVED** [review]

File: examples/consistency_distillation/train_lcm_distill_lora_sd_wds.py [hunk 2]
```diff
@@ -360,18 +360,42 @@ def scalings_for_boundary_conditions(timestep, sigma_data=0.5, timestep_scaling=
...
+        raise ValueError(
+            f"Prediction type {prediction_type} is not supported; currently, `epsilon`, `sample`, and `v_prediction`"
+            f" are supported."
+        )
 
     return pred_x_0
 
 
+# Based on step 4 in DDIMScheduler.step
+def predicted_source_noise(model_output, timesteps, sample, prediction_type, alphas, sigmas):
```
Comment (patil-suraj) [inline examples/consistency_distillation/train_lcm_distill_lora_sd_wds.py:381]: what does `source` refer to here ?

File: examples/consistency_distillation/train_lcm_distill_lora_sd_wds.py [hunk 3]
```diff
@@ -835,16 +859,17 @@ def main(args):
...
     sigma_schedule = torch.sqrt(1 - noise_scheduler.alphas_cumprod)
+    # Initialize the DDIM ODE solver for distillation.
     solver = DDIMSolver(
         noise_scheduler.alphas_cumprod.numpy(),
         timesteps=noise_scheduler.config.num_train_timesteps,
         ddim_timesteps=args.num_ddim_timesteps,
     )
 
-    # 2. Load tokenizers from SD-XL checkpoint.
+    # 2. Load tokenizers from SD-1.5 checkpoint.
```
Comment (patil-suraj) [inline examples/consistency_distillation/train_lcm_distill_lora_sd_wds.py:872]: this script should also work for SD2.1 , so maybe let's mention SD1.x and SD2.x

File: examples/consistency_distillation/train_lcm_distill_lora_sd_wds.py [hunk 4]
```diff
@@ -855,14 +880,14 @@ def main(args):
         args.pretrained_teacher_model, subfolder="text_encoder", revision=args.teacher_revision
     )
 
-    # 4. Load VAE from SD-XL checkpoint (or more stable VAE)
+    # 4. Load VAE from SD-1.5 checkpoint
```
Comment (patil-suraj) [inline examples/consistency_distillation/train_lcm_distill_lora_sd_wds.py:883]: same comment as above

File: examples/consistency_distillation/train_lcm_distill_lora_sd_wds.py [hunk 5]
```diff
@@ -855,14 +880,14 @@ def main(args):
...
     vae = AutoencoderKL.from_pretrained(
         args.pretrained_teacher_model,
         subfolder="vae",
         revision=args.teacher_revision,
     )
 
-    # 5. Load teacher U-Net from SD-XL checkpoint
+    # 5. Load teacher U-Net from SD-1.5 checkpoint
```
Comment (patil-suraj) [inline examples/consistency_distillation/train_lcm_distill_lora_sd_wds.py:890]: same comment as above

File: examples/consistency_distillation/train_lcm_distill_lora_sd_wds.py [hunk 6]
```diff
@@ -872,7 +897,7 @@ def main(args):
     text_encoder.requires_grad_(False)
     teacher_unet.requires_grad_(False)
 
-    # 7. Create online (`unet`) student U-Nets.
+    # 7. Create online (`unet`) student U-Net.
```
Comment (patil-suraj) [inline examples/consistency_distillation/train_lcm_distill_lora_sd_wds.py:900]:
```suggestion
    # 7. Create online  student U-Net.
```

File: examples/consistency_distillation/train_lcm_distill_lora_sd_wds.py [hunk 7]
```diff
@@ -1141,36 +1171,38 @@ def compute_embeddings(prompt_batch, proportion_empty_prompts, text_encoder, tok
...
                 latents = latents.to(weight_dtype)
 
-                # Sample noise that we'll add to the latents
+                # 2. Sample noise that we'll add to the latents
                 noise = torch.randn_like(latents)
                 bsz = latents.shape[0]
 
-                # Sample a random timestep for each image t_n ~ U[0, N - k - 1] without bias.
+                # 3. Sample a random timestep for each image t_n from the ODE solver timesteps without bias.
+                # For the DDIM solver, the timestep schedule is [T - 1, T - k - 1, T - 2 * k - 1, ...]
```
Comment (patil-suraj) [inline examples/consistency_distillation/train_lcm_distill_lora_sd_wds.py:1179]: really nice comments !

File: examples/consistency_distillation/train_lcm_distill_lora_sd_wds.py [hunk 8]
```diff
@@ -1141,36 +1171,38 @@ def compute_embeddings(prompt_batch, proportion_empty_prompts, text_encoder, tok
...
 
-                # 20.4.4. Get boundary scalings for start_timesteps and (end) timesteps.
+                # 4. Get boundary scalings for start_timesteps and (end) timesteps.
                 c_skip_start, c_out_start = scalings_for_boundary_conditions(start_timesteps)
                 c_skip_start, c_out_start = [append_dims(x, latents.ndim) for x in [c_skip_start, c_out_start]]
                 c_skip, c_out = scalings_for_boundary_conditions(timesteps)
                 c_skip, c_out = [append_dims(x, latents.ndim) for x in [c_skip, c_out]]
 
-                # 20.4.5. Add noise to the latents according to the noise magnitude at each timestep
+                # 5. Add noise to the latents according to the noise magnitude at each timestep
```
Comment (patil-suraj) [inline examples/consistency_distillation/train_lcm_distill_lora_sd_wds.py:1192]: Let's init the noise here, just before we add it, rather than up above. This would be more readable IMO.

File: examples/consistency_distillation/train_lcm_distill_lora_sd_wds.py [hunk 9]
```diff
@@ -1208,8 +1242,16 @@ def compute_embeddings(prompt_batch, proportion_empty_prompts, text_encoder, tok
                             alpha_schedule,
                             sigma_schedule,
                         )
+                        cond_pred_noise = predicted_source_noise(
```
Comment (patil-suraj) [inline examples/consistency_distillation/train_lcm_distill_lora_sd_wds.py:1245]: I would call the metho something like `get_predicted_noise`

File: examples/consistency_distillation/train_lcm_distill_lora_sdxl_wds.py [hunk 10]
```diff
@@ -347,18 +347,42 @@ def scalings_for_boundary_conditions(timestep, sigma_data=0.5, timestep_scaling=
...
-        raise ValueError(f"Prediction type {prediction_type} currently not supported.")
+        raise ValueError(
+            f"Prediction type {prediction_type} is not supported; currently, `epsilon`, `sample`, and `v_prediction`"
+            f" are supported."
+        )
 
     return pred_x_0
 
 
+# Based on step 4 in DDIMScheduler.step
```
Comment (patil-suraj) [inline examples/consistency_distillation/train_lcm_distill_lora_sdxl_wds.py:367]: same comment as above regarding the name

File: examples/consistency_distillation/train_lcm_distill_lora_sdxl_wds.py [hunk 11]
```diff
@@ -886,7 +911,7 @@ def main(args):
     text_encoder_two.requires_grad_(False)
     teacher_unet.requires_grad_(False)
 
-    # 7. Create online (`unet`) student U-Nets.
+    # 7. Create online (`unet`) student U-Net.
```
Comment (patil-suraj) [inline examples/consistency_distillation/train_lcm_distill_lora_sdxl_wds.py:914]:
```suggestion
    # 7. Create student U-Net.
```

File: examples/consistency_distillation/train_lcm_distill_lora_sdxl_wds.py [hunk 12]
```diff
@@ -1197,36 +1224,38 @@ def compute_embeddings(
...
 
-                # 20.4.4. Get boundary scalings for start_timesteps and (end) timesteps.
+                # 4. Get boundary scalings for start_timesteps and (end) timesteps.
                 c_skip_start, c_out_start = scalings_for_boundary_conditions(start_timesteps)
                 c_skip_start, c_out_start = [append_dims(x, latents.ndim) for x in [c_skip_start, c_out_start]]
                 c_skip, c_out = scalings_for_boundary_conditions(timesteps)
                 c_skip, c_out = [append_dims(x, latents.ndim) for x in [c_skip, c_out]]
 
-                # 20.4.5. Add noise to the latents according to the noise magnitude at each timestep
+                # 5. Add noise to the latents according to the noise magnitude at each timestep
```
Comment (patil-suraj) [inline examples/consistency_distillation/train_lcm_distill_lora_sdxl_wds.py:1245]: same comment as above, re noise

File: examples/consistency_distillation/train_lcm_distill_sd_wds.py [hunk 13]
```diff
@@ -337,18 +337,42 @@ def scalings_for_boundary_conditions(timestep, sigma_data=0.5, timestep_scaling=
...
-        raise ValueError(f"Prediction type {prediction_type} currently not supported.")
+        raise ValueError(
+            f"Prediction type {prediction_type} is not supported; currently, `epsilon`, `sample`, and `v_prediction`"
+            f" are supported."
+        )
 
     return pred_x_0
 
 
+# Based on step 4 in DDIMScheduler.step
```
Comment (patil-suraj) [inline examples/consistency_distillation/train_lcm_distill_sd_wds.py:357]: same comment as above re name

File: examples/consistency_distillation/train_lcm_distill_sd_wds.py [hunk 14]
```diff
@@ -823,16 +847,17 @@ def main(args):
...
     sigma_schedule = torch.sqrt(1 - noise_scheduler.alphas_cumprod)
+    # Initialize the DDIM ODE solver for distillation.
     solver = DDIMSolver(
         noise_scheduler.alphas_cumprod.numpy(),
         timesteps=noise_scheduler.config.num_train_timesteps,
         ddim_timesteps=args.num_ddim_timesteps,
     )
 
-    # 2. Load tokenizers from SD-XL checkpoint.
+    # 2. Load tokenizers from SD-1.5 checkpoint.
```
Comment (patil-suraj) [inline examples/consistency_distillation/train_lcm_distill_sd_wds.py:860]: Should work for both SD1.x and SD2.x checkpoints.

File: examples/consistency_distillation/train_lcm_distill_sd_wds.py [hunk 15]
```diff
@@ -860,7 +885,7 @@ def main(args):
     text_encoder.requires_grad_(False)
     teacher_unet.requires_grad_(False)
 
-    # 8. Create online (`unet`) student U-Nets. This will be updated by the optimizer (e.g. via backpropagation.)
+    # 7. Create online (`unet`) student U-Net. This will be updated by the optimizer (e.g. via backpropagation.)
```
Comment (patil-suraj) [inline examples/consistency_distillation/train_lcm_distill_sd_wds.py:888]:
```suggestion
    # 7. Create student U-Net. This will be updated by the optimizer (e.g. via backpropagation.)
```

File: examples/consistency_distillation/train_lcm_distill_sd_wds.py [hunk 16]
```diff
@@ -1124,39 +1153,40 @@ def compute_embeddings(prompt_batch, proportion_empty_prompts, text_encoder, tok
...
 
-                # 20.4.4. Get boundary scalings for start_timesteps and (end) timesteps.
+                # 4. Get boundary scalings for start_timesteps and (end) timesteps.
                 c_skip_start, c_out_start = scalings_for_boundary_conditions(start_timesteps)
                 c_skip_start, c_out_start = [append_dims(x, latents.ndim) for x in [c_skip_start, c_out_start]]
                 c_skip, c_out = scalings_for_boundary_conditions(timesteps)
                 c_skip, c_out = [append_dims(x, latents.ndim) for x in [c_skip, c_out]]
 
-                # 20.4.5. Add noise to the latents according to the noise magnitude at each timestep
+                # 5. Add noise to the latents according to the noise magnitude at each timestep
```
Comment (patil-suraj) [inline examples/consistency_distillation/train_lcm_distill_sd_wds.py:1174]: same comment re noise.

**Review by patil-suraj — APPROVED** [review]
Really nice and explanatory comments, great job!
Looks good to me, just left some nits.

File: examples/consistency_distillation/train_lcm_distill_lora_sd_wds.py (hunk 2 above)
Comment (dg845) [reply examples/consistency_distillation/train_lcm_distill_lora_sd_wds.py:381]: Here `source` refers to the source noise $`\epsilon`$ ($`\epsilon_t`$ in Equation (5) of the [DDIM paper](https://arxiv.org/pdf/2010.02502.pdf)) that the denoising model tries to predict when `prediction_type == "epsilon"`.

Would it be more clear is this was named something like `predicted_epsilon` (which uses terminology closer to that of `DDIMScheduler`)?

Comment (dg845) [reply examples/consistency_distillation/train_lcm_distill_lora_sd_wds.py:381]: Saw https://github.com/huggingface/diffusers/pull/6145#discussion_r1427694793 below, `get_predicted_noise` works for me
